### PR TITLE
feat(compaction):  dynamically determine the type filter_key_extractor_manager

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/filter_key_extractor.rs [hunk 1]
```diff
@@ -376,11 +375,52 @@ impl FilterKeyExtractorManager {
...
+
+impl FilterKeyExtractorBuilder {
+    pub fn new(id_to_table: HashMap<u32, Table>) -> Self {
+        Self { id_to_table }
+    }
+
+    fn acquire(&self, table_id_set: HashSet<u32>) -> HummockResult<FilterKeyExtractorImpl> {
+        let mut multi_filter_key_extractor = MultiFilterKeyExtractor::default();
+        for table_id in table_id_set {
+            if let Some(table) = self.id_to_table.get(&table_id) {
```
Comment (hzxa21) [inline src/storage/src/filter_key_extractor.rs:418]: What happen if some tables in the `table_id_set` are absent in `id_to_table`? Should we return an error for that?

Comment (wcy-fdu) [reply src/storage/src/filter_key_extractor.rs:418]: Let's return error as a backstop, and proxy should guarantee that this will not happen.

File: src/storage/hummock_test/src/sync_point_tests.rs [hunk 2]
```diff
@@ -251,7 +251,8 @@ pub async fn compact_once(
     compact_task.compaction_filter_mask = compaction_filter_flag.bits();
     // 3. compact
     let (_tx, rx) = tokio::sync::oneshot::channel();
-    let (mut result_task, task_stats) = compact(compact_ctx, compact_task.clone(), rx).await;
+    let (mut result_task, task_stats) =
+        compact(compact_ctx.clone(), compact_task.clone(), rx).await;
```
Comment (hzxa21) [inline src/storage/hummock_test/src/sync_point_tests.rs:255]: nits: unnecessary change

File: src/storage/src/hummock/compactor/mod.rs [hunk 3]
```diff
@@ -503,7 +503,7 @@ pub fn start_compactor(compactor_context: Arc<CompactorContext>) -> (JoinHandle<
                                         let (tx, rx) = tokio::sync::oneshot::channel();
                                         let task_id = compact_task.task_id;
                                         shutdown.lock().unwrap().insert(task_id, tx);
-                                        let (compact_task, table_stats) = compactor_runner::compact(context, compact_task, rx).await;
+                                        let (compact_task, table_stats)=  compactor_runner::compact(context.clone(), compact_task, rx).await;
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/mod.rs:506]: fmt

File: src/storage/src/filter_key_extractor.rs [hunk 4]
```diff
@@ -376,15 +375,60 @@ impl FilterKeyExtractorManager {
...
     /// `table_id` does not util version update (notify), and retry to get
+    async fn acquire(&self, table_id_set: HashSet<u32>) -> HummockResult<FilterKeyExtractorImpl> {
+        self.inner.acquire(table_id_set).await
+    }
+}
+
+#[derive(Clone)]
+pub enum FilterKeyExtractorManager {
+    RpcFilterKeyExtractorManager(Arc<RpcFilterKeyExtractorManager>),
+    StaticFilterKeyExtractorManager(StaticFilterKeyExtractorManager),
```
Comment (hzxa21) [inline src/storage/src/filter_key_extractor.rs:386]: We should also wrap `StaticFilterKeyExtractorManager` in Arc. Otherwise, the clone will be come heavy since the whole HashMap with all table catalogs will be copied.

File: src/storage/src/hummock/mod.rs [hunk 5]
```diff
@@ -257,7 +260,12 @@ impl HummockStorage {
     }
 
     pub fn filter_key_extractor_manager(&self) -> &FilterKeyExtractorManagerRef {
```
Comment (hzxa21) [inline src/storage/src/hummock/mod.rs:262]: I think it is better to return the `FilterKeyExtractorManager` enum type here.

**Review by hzxa21 — APPROVED** [review]
Rest LGTM.

File: src/storage/src/hummock/compactor/mod.rs (hunk 3 above)
Comment (wcy-fdu) [reply src/storage/src/hummock/compactor/mod.rs:506]: wired,` cargo fmt i`s invalidated in async move.

File: src/storage/src/filter_key_extractor.rs (hunk 4 above)
Comment (wcy-fdu) [reply src/storage/src/filter_key_extractor.rs:386]: Nice catch!
